Added generator for ISO11649 Creditor Reference

kohlerdominik opened the pull request:
I found this missing and had to do some investigation before i found out how to work with it.

In my opinion, it makes sense to add a generator for this, working the same way as the `QrPaymentReferenceGenerator`. I think that's what you're looking for as a developer.

Looking at the QrPaymentReferenceGenerator:
Maybe I'm not understanding it, but as I see there a Validator in place, which is used in the `doGenerate()`-method. However, `getViolations()` seems inaccessible, as the instasnce is never exposed.
Of curse it's possible to get a new instance with `new`, but then it's not possible to set the properties.

Because of this I used a different approach for the creation and the facade in RfCreditorReferenceGenerator. But I'm open for inputs.

sprain commented on src/Reference/QrPaymentReferenceGenerator.php:
```diff
@@ -24,9 +25,9 @@ public static function generate(?string $customerIdentificationNumber, string $r
         if (null !== $customerIdentificationNumber) {
-            $qrPaymentReferenceGenerator->customerIdentificationNumber = $qrPaymentReferenceGenerator->removeWhitespace($customerIdentificationNumber);
+            $qrPaymentReferenceGenerator->customerIdentificationNumber = StringModifier::stripWhitespace($customerIdentificationNumber);
         }
-        $qrPaymentReferenceGenerator->referenceNumber = $qrPaymentReferenceGenerator->removeWhitespace($referenceNumber);
+        $qrPaymentReferenceGenerator->referenceNumber = StringModifier::stripWhitespace($referenceNumber);
```
Thanks! That was a sleeping todo of mine :)

sprain commented:
> In my opinion, it makes sense to add a generator for this, working the same way as the QrPaymentReferenceGenerator. I think that's what you're looking for as a developer.

This is a very good idea. I will be happy to have a look at this pr :)

sprain commented:
> Maybe I'm not understanding it, but as I see there a Validator in place, which is used in the doGenerate()-method. However, getViolations() seems inaccessible, as the instasnce is never exposed.

It's available in `SelfValidatableTrait`.
You can do `$qrPaymentReferenceGenerator->getViolations()`

kohlerdominik commented:
Fixed broken tests in `--prefer-lowest` and fixed cs.

It does not make sense to allow multiple versions of phpunit, as as dev-dependency its only used in the context of package development anyway. I upgraded to the latest 7.*-version to support `assertStringContainsStringIgnoringCase`-assertation. However, maybe with the compatibility tests with PHP8, we should also upgrade to a supported version of phpunit.

For symfony/validator, I added `NotBlank`-assertation. 
But due inconsistent behaviour `Length`-assertation between v3.4 (`failed:[NotBlank]` and v5.0 [`failed[NotBlank, Length]`) it's probably better to drop the legacy version support and remove `NotBlank` in combination with `Length at all.`

sprain commented:
> It does not make sense to allow multiple versions of phpunit, as as dev-dependency its only used in the context of package development anyway. I upgraded to the latest 7.*-version to support assertStringContainsStringIgnoringCase-assertation.

Great, thanks!

> However, maybe with the compatibility tests with PHP8, we should also upgrade to a supported version of phpunit.

Indeed this is an open task. I'll also create an issue for it.

> But due inconsistent behaviour Length-assertation between v3.4 (failed:[NotBlank] and v5.0 [failed[NotBlank, Length]) it's probably better to drop the legacy version support and remove NotBlank in combination with Length at all.

I am not sure what you mean here. Could you ellaborate?
Symfony 3.4 is still maintained for quite a while and also still used in many projects. Therefore I do not want to drop support for those dependencies yet.

sdespont commented:
> Symfony 3.4 is still maintained for quite a while and also still used in many projects. Therefore I do not want to drop support for those dependencies yet.

Agree with that. It is better to support both in code using conditional statement if there is incompatibility.

kohlerdominik commented:
> I am not sure what you mean here. Could you ellaborate?
> Symfony 3.4 is still maintained for quite a while and also still used in many projects. Therefore I do not want to drop support for those dependencies yet.

**Note:** it's just a suggestion. For me personally or my projects it doesn't matter at all. I'm happy with whatever decision of yours.

To express why i made the suggestion, however:
The `Length`-validation does not behave consistent.
```
Assume this configuration:
Length(['min'=>1,  'max'=>3])

LengthValidator: v3.4
'': valid
'12': valid
'1234': invalid

LengthValidator: v5.0
'': invalid
'12': valid
'1234': invalid
```

So, in v3.4 a zero-length-string is valid, while in 5.0 it's invalid. Therefore, for backwards-compatibility with v3.4 it's necessary to add NotBlank-validator to make the validator work. This Workaround if perfectly fine, BUT it will result in inconsist behaviour:

```
Length(['min'=>1,  'max'=>3]), NotBlank()
v3.4: Violations: [0 => 'NotBlank']
v5.0: Violations: [0 => 'Length', 1 => 'NotBlank']
```

So your application can have one or two violations, depending on the Symfony-version. It's no big deal, but it's incosistent.

sprain commented:
Thanks a lot for your work and explanations!
I am quite busy at the time, but will get back to this and have a look asap.

sprain commented:
Just an fyi: I have not forgotten about this and will get back to it.

sprain commented:
Thanks a lot for this! It has now been merged. Release will follow when tests pass on master.
I also adjusted QrPaymentReferenceGenerator to match the structure of your approach (see https://github.com/sprain/php-swiss-qr-bill/commit/e503ad1c4d3a26f023b40dbc9627ea6355899948). It's better because violations now actually are accessible, if desired.
